### PR TITLE
mon/OSDMonitor: remove zeroed new_state updates

### PR DESCRIPTION
**Pull request by liewegas**
Fixes: http://tracker.ceph.com/issues/20751

### CODE REVIEW

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -942,6 +942,18 @@ void OSDMonitor::encode_pending(MonitorDBStore::TransactionRef t)
   }
   mapping_job.reset();
 
+  // ensure we don't have blank new_state updates.  these are interrpeted as
+  // CEPH_OSD_UP (and almost certainly not what we want!).
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: i think, if `new_state` is `0`, then it has no effect on the new osdmap after it is applied to the old one.

**Comment by liewegas** [comment]
https://github.com/ceph/ceph/blob/master/src/osd/OSDMap.cc#L1671

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-07-24_03:44:49-rados-wip-sage-testing-distro-basic-smithi/

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -942,6 +942,18 @@ void OSDMonitor::encode_pending(MonitorDBStore::TransactionRef t)
...
+  auto p = pending_inc.new_state.begin();
+  while (p != pending_inc.new_state.end()) {
+    if (p->second == 0) {
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: @liewegas but how come can we add a `0` new_state mask? this happens only if we have a new `0` state mask in the incoming MOSDFull. the only case i can think of is #16581.

Comment (liewegas) [inline src/mon/OSDMonitor.cc]: Ah, yeah, that was probably it!

**Pull request merged**